eva/readers/decord_reader.py

```diff
@@ -70,50 +88,50 @@ def _read(self) -> Iterator[Dict]:
         logger.debug("Reading frames")
 
         if self._sampling_type == IFRAMES:
-            iframes = video.get_key_indices()
+            assert not self._read_audio, "Cannot use IFRAMES with audio streams"
+            iframes = reader.get_key_indices()
             idx = 0
             for begin, end in range_list:
                 while idx < len(iframes) and iframes[idx] < begin:
                     idx += self._sampling_rate
 
                 while idx < len(iframes) and iframes[idx] <= end:
                     frame_id = iframes[idx]
-                    frame = video[frame_id].asnumpy()
                     idx += self._sampling_rate
-                    if frame is not None:
-                        yield {
-                            "id": frame_id,
-                            "data": frame,
-                            "seconds": round(video.get_frame_timestamp(frame_id)[0], 2),
-                        }
-                    else:
-                        break
+                    yield get_frame(frame_id, reader)
+
         elif self._sampling_rate == 1:
             for begin, end in range_list:
                 frame_id = begin
                 while frame_id <= end:
-                    frame = video[frame_id].asnumpy()
-                    if frame is not None:
-                        yield {
-                            "id": frame_id,
-                            "data": frame,
-                            "seconds": round(video.get_frame_timestamp(frame_id)[0], 2),
-                        }
-                    else:
-                        break
+                    yield get_frame(frame_id, reader)
                     frame_id += 1
         else:
             for begin, end in range_list:
                 # align begin with sampling rate
                 if begin % self._sampling_rate:
                     begin += self._sampling_rate - (begin % self._sampling_rate)
                 for frame_id in range(begin, end + 1, self._sampling_rate):
-                    frame = video[frame_id].asnumpy()
-                    if frame is not None:
-                        yield {
-                            "id": frame_id,
-                            "data": frame,
-                            "seconds": round(video.get_frame_timestamp(frame_id)[0], 2),
-                        }
-                    else:
-                        break
+                    yield get_frame(frame_id, reader)
+
+    def __get_video_frame(self, frame_id, reader):
+        frame_video = reader[frame_id]
+        frame_video = frame_video.asnumpy()
+        timestamp = reader.get_frame_timestamp(frame_id)[0]
+
+        return {
+            VideoColumnName.id.name: frame_id,
+            VideoColumnName.data.name: frame_video,
+            VideoColumnName.seconds.name: round(timestamp, 2),
+        }
+
+    def __get_audio_frame(self, frame_id, reader):
+        frame_audio, _ = reader[frame_id]
```

Comment:
So, we are going with audio per frame design. Will this work for whisper? If yes, we are good. We can worry about the sampling rate in the next PR when we add whisper.

Reply:
Yes, I verified that it works with whisper